### PR TITLE
RSDK-251 Add support for capsule geometry types

### PR DESCRIPTION
**Pull request by biotinker**
This adds capsules as a geometry type, supporting efficient collisions with all other geometry types.

Along with this is an initial implementation of meshes- not supported as a geometry type, simply as a backend representation. This is due to a box-capsule collision implementation being taken from here: https://github.com/google/brax/blob/7eaa16b4bf446b117b538dbe9c9401f97cf4afa2/brax/physics/colliders.py

The separating axis theorum is instead used here for the capsules when doing collisions as it is faster. Perhaps we should PR against brax?

### CODE REVIEW

File: spatialmath/geometry.go [hunk 1]
```diff
@@ -15,6 +15,8 @@ type Geometry interface {
 	Transform(Pose) Geometry
 	ToProtobuf() *commonpb.Geometry
 	CollidesWith(Geometry) (bool, error)
+	// If DistanceFrom is negative, it represents the penetration depth of the two geometries, which are in collision.
+	// Penetration depth magnitude is defined as the minimum translation which would result in the geometries not colliding.
```
Comment (raybjork) [inline spatialmath/geometry.go:19]: Could we add the disclaimer on this function that for certain entity pairs we do a conservative estimate?  I feel like we should be advertising that this is something of an experimental feature at the moment

File: spatialmath/sphere.go [hunk 2]
```diff
@@ -116,6 +121,9 @@ func (s *sphere) EncompassedBy(g Geometry) (bool, error) {
 	if other, ok := g.(*sphere); ok {
 		return sphereInSphere(s, other), nil
 	}
+	if other, ok := g.(*capsule); ok {
+		return sphereInCapsule(s, other), nil
```
Comment (raybjork) [inline spatialmath/sphere.go:125]: I think you will also need to add checks like this to the point.go file

File: spatialmath/geometry_test.go [hunk 3]
```diff
@@ -539,3 +548,338 @@ func TestSphereVsPointEncompassed(t *testing.T) {
 	}
 	testGeometryEncompassed(t, cases)
 }
+
+func TestCapsuleVsBoxCollision(t *testing.T) {
```
Comment (raybjork) [inline spatialmath/geometry_test.go:552]: Will come back and look at this file in a second pass later

File: spatialmath/geometry_utils.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+const floatEpsilon = 1e-6
+
+// This file incorporates work covered by the Brax project -- https://github.com/google/brax/blob/main/LICENSE.
+// Copyright 2021 The Brax Authors, which is licensed under the Apache License Version 2.0 (the “License”).
+// You may obtain a copy of the license at http://www.apache.org/licenses/LICENSE-2.0.
+
+// DistToLineSegment takes a line segment defined by pt1 and pt2, plus some query point, and returns the cartesian distance from the
+// query point to the closest point on the line segment.
+func DistToLineSegment(pt1, pt2, query r3.Vector) float64 {
```
Comment (raybjork) [inline spatialmath/geometry_utils.go:15]: A lot of the code in this file feels familiar to me.  Is there a chance it is duplicating other code?

**Review by raybjork — CHANGES_REQUESTED** [review]
Looking good so far.  I had a few questions mostly related to the algorithm were using for box-cylinder collisions (and the necessity of meshes in general).  I also left a couple comments about what I haven't looked at yet and will get back to in the next pass

File: spatialmath/box.go [hunk 5]
```diff
@@ -193,8 +186,52 @@ func (b *box) penetrationDepth(pt r3.Vector) float64 {
...
+func (b *box) toMesh() *mesh {
+	if b.triangles != nil {
+		return b.triangles
+	}
+	m := &mesh{pose: b.pose}
+	triangles := make([]*triangle, 0, 12)
+	verts := b.vertices()
+
+	// Lists of adjacent face corners not including the opposing ones
+	// TODO(pl): Do this in a way that doesn't require a bunch of magic numbers or duplicated code
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/box.go:213]: Agree. Probably the triples is the way to go (or at least indexing the triples)

File: spatialmath/capsule.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+	"go.viam.com/rdk/utils"
+)
+
+// capsule is a collision geometry that represents a capsule, it has a pose and a radius that fully define it.
+//
+//	 __________________
+//	/                  \
+//
+// x|  |--------------|  |x
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:19]: This was aligned; the linter did this. I'll have to figure out how to make it not reformat comments.

File: spatialmath/capsule.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+	// this is the pose of one end of the capsule. The full capsule extends `length` mm outwards in the direction of
+	// the pose's orientation
+	pose   Pose
+	radius float64
+	length float64 // total length of the capsule, tip to tip
+	label  string
+
+	// These values are generated at geometry creation time and should not be altered by hand
+	// They are stoed here because they are useful and expensive to calculate
+	segA r3.Vector // Proximal endpoint of capsule line segment. First point from `pose` to be surrounded by `radius` of capsule
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:34]: Optimization. Most calculations involving capsules involve segA and segB, so it makes sense to compute them up-front.

File: spatialmath/capsule.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+	segB r3.Vector // Distal endpoint of capsule line segment. Most distal point to be surrounded by `radius` of capsule
+}
+
+// NewCapsule instantiates a new capsule Geometry.
+func NewCapsule(offset Pose, radius, length float64, label string) (Geometry, error) {
+	if radius < 0 {
+		return nil, newBadGeometryDimensionsError(&capsule{})
+	}
+	if length == radius*2 || length == 0 {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:43]: The convenience here is allowing spheres and capsules to be e.g. inferred from proto with the same code.

If we allow `0 <= length <= 2*r`, then the user can pass in potentially meaningful values that will be silently ignored. `0` is the only value in that range where the user is getting exactly what they asked for.

File: spatialmath/capsule.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+	}
+	if length < radius*2 {
+		return nil, newBadCapsuleLengthError(length, radius)
+	}
+	return precalcCapsule(offset, radius, length, label), nil
+}
+
+// Will precalculate the linear endpoints for a capsule.
+func precalcCapsule(offset Pose, radius, length float64, label string) Geometry {
+	segA := Compose(offset, NewPoseFromPoint(r3.Vector{0, 0, radius})).Point()
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:54]: Agree that inconsistency should be addressed.

I will do a quick user survey to see what makes the most intuitive sense.

File: spatialmath/capsule.go [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+	return c.pose
+}
+
+// AlmostEqual compares the capsule with another geometry and checks if they are equivalent.
+func (c *capsule) AlmostEqual(g Geometry) bool {
+	other, ok := g.(*capsule)
+	if !ok {
+		return false
+	}
+	return PoseAlmostEqual(c.pose, other.pose) && utils.Float64AlmostEqual(c.radius, other.radius, 1e-8)
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:102]: Good catch

File: spatialmath/capsule.go [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// Transform premultiplies the capsule pose with a transform, allowing the capsule to be moved in space.
+func (c *capsule) Transform(toPremultiply Pose) Geometry {
+	return precalcCapsule(Compose(toPremultiply, c.pose), c.radius, c.length, c.label)
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:107]: I'm fine with that

File: spatialmath/capsule.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+	if other, ok := g.(*sphere); ok {
+		return capsuleInSphere(c, other), nil
+	}
+	if _, ok := g.(*point); ok {
+		return false, nil
+	}
+	return true, newCollisionTypeUnsupportedError(c, g)
+}
+
+// ToPointCloud converts a capsule geometry into []r3.Vector. This method takes one argument which determines
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:169]: Linter missed this because it's a method whose comment is on the interface definition.

I'll need to fix this in `Sphere` as well since that's where a lot of this was copied from.

File: spatialmath/capsule.go [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+// how many points should like on the capsule's surface. If the argument is set to 0. we automatically
+// substitute the value with defaultTotalSpherePoints.
+func (c *capsule) ToPoints(resolution float64) []r3.Vector {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:172]: This should be able to start by calling sphere.ToPoints() honestly, I'll do that.

File: spatialmath/geometry.go (hunk 1 above)
Comment (biotinker) [reply spatialmath/geometry.go:19]: Yes, will do

File: spatialmath/capsule.go [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+	return DistToLineSegment(c.segA, c.segB, other.pose.Point()) - (c.radius + other.radius)
+}
+
+func capsuleVsCapsuleDistance(c, other *capsule) float64 {
+	return SegmentDistanceToSegment(c.segA, c.segB, other.segA, other.segB) - (c.radius + other.radius)
+}
+
+func capsuleVsBoxDistance(c *capsule, other *box) float64 {
+	// Capsules are not polyhedrons, so separating axis test is not ideal here.
+	meshDist := capsuleVsMeshDistance(c, other.toMesh())
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:232]: I should probably swap this over to the Separating Axis Theorum as well.

There were *no* examples I could find of using the SAT for capsule vs box, and implementations I did find tended to convert capsules into triangles and do a mesh collision, hence why meshes are partially implemented in this PR. Since that's how it's commonly done, I began with the assumption it would be faster than the SAT.

Turns out it's slower than the SAT by quite a bit, so once I could compare I swapped back over. I'll swap this as well.

File: spatialmath/geometry_utils.go (hunk 4 above)
Comment (biotinker) [reply spatialmath/geometry_utils.go:15]: This was meant to be moved over from motionplan/constraints.go, but it appears I forgot to actually commit that change.

File: spatialmath/mesh.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+import (
+	"github.com/golang/geo/r3"
+)
+
+// This file incorporates work covered by the Brax project -- https://github.com/google/brax/blob/main/LICENSE.
+// Copyright 2021 The Brax Authors, which is licensed under the Apache License Version 2.0 (the “License”).
+// You may obtain a copy of the license at http://www.apache.org/licenses/LICENSE-2.0.
+
+// mesh is a collision geometry that represents a set of triangles that represent a mesh.
+type mesh struct {
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/mesh.go:12]: I tested the math correctness pretty extensively so I'm confident in it, but feel free!

File: spatialmath/sphere.go (hunk 2 above)
Comment (biotinker) [reply spatialmath/sphere.go:125]: Ack, good catch

File: spatialmath/capsule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (raybjork) [reply spatialmath/capsule.go:43]: I don't follow this.  Proto geometries are strongly typed so no need for inference.  Config geometry is inferred but if any L is specified we know we can be using a capsule.  Can you give an example of how this would be broken?

File: spatialmath/box.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/box.go:213]: Actually, I'm going to remove this function entirely, since we're moving away from boxes-as-mesh for capsule collisions.

File: spatialmath/capsule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:41]: I was referring to configs, not raw protos, apologies; I misspoke. The ode in question is here:

https://github.com/viamrobotics/rdk/blob/140edb771726c511cecd4e9b68a7f2c78115eff4/spatialmath/geometry.go#L105

If L is zero, the we return a sphere, which is correct.

If L is 2, but radius is 1.5, then that's an invalid capsule, as length would have to be greater than 3. We shouldn't coerce to a sphere in this case, because it's unknown if the user intended a capsule of length 5, or a sphere of diameter 3.

In that case we should error to the user.

File: spatialmath/capsule.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:52]: Thought on this more.

Most implementations of capsule out there are radius, base, tip, hence why ours is defined off of base.

But this inconsistency within RDK would not be worth that. I'll swap this over to be midpoint.

File: spatialmath/capsule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (raybjork) [reply spatialmath/capsule.go:41]: Ah ok I follow now.  I'm not sure I entirely agree that reporting that as an error is the best course of action, but I can get behind it.

[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:41]: If you prefer I can make it so that length == 0 will also be an error; that way we're never assuming the user is shortcutting.

File: spatialmath/capsule.go [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+		Label: c.label,
+	}
+}
+
+// CollidesWith checks if the given capsule collides with the given geometry and returns true if it does.
+func (c *capsule) CollidesWith(g Geometry) (bool, error) {
+	if other, ok := g.(*box); ok {
+		return capsuleVsBoxCollision(c, other), nil
+	}
+	dist, err := c.DistanceFrom(g)
```
Comment (raybjork) [inline spatialmath/capsule.go:129]: For the other geometry types I found that it was not efficient to return the distance if we want simple collision, so I wrote two methods, one for each, for optimality's sake.  Do you think theres benefit to be had here or no?

File: spatialmath/capsule.go [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+	return true, newCollisionTypeUnsupportedError(c, g)
+}
+
+// ToPoints converts a capsule geometry into []r3.Vector. This method takes one argument which determines
+// how many points should like on the capsule's surface. If the argument is set to 0. we automatically
+// substitute the value with defaultTotalSpherePoints.
+func (c *capsule) ToPoints(resolution float64) []r3.Vector {
+	s := &sphere{pose: NewZeroPose(), radius: c.radius}
+	vecList := s.ToPoints(resolution)
+	// move points to be correctly located on capsule endcaps
```
Comment (raybjork) [inline spatialmath/capsule.go:175]: this is cool

File: spatialmath/capsule.go [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+	return pointVsBoxDistance(c.segA, b) <= -c.radius && pointVsBoxDistance(c.segB, b) <= -c.radius
+}
+
+// capsuleInSphere returns a bool describing if the given capsule is fully encompassed by the given sphere.
+func capsuleInSphere(c *capsule, s *sphere) bool {
+	return c.segA.Sub(s.pose.Point()).Norm()+c.radius <= s.radius && c.segB.Sub(s.pose.Point()).Norm()+c.radius <= s.radius
+}
+
+// capsuleBoxSeparatingAxis returns immediately as soon as any result is found indicating that the two objects are not in collision.
+func capsuleBoxSeparatingAxis(a *capsule, b *box) bool {
```
Comment (raybjork) [inline spatialmath/capsule.go:268]: this should be `c *capsule` , matching the rest of the code

File: spatialmath/capsule.go [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+
+	// check if there is a distance between bounding spheres to potentially exit early
+	if centerDist.Norm()-((a.length/2)+b.boundingSphereR) > CollisionBuffer {
+		return false
+	}
+	rmA := a.pose.Orientation().RotationMatrix()
+	rmB := b.pose.Orientation().RotationMatrix()
+
+	segLen := (a.length - 2*a.radius) / 2
+	capVec := rmA.Row(2).Mul(segLen)
```
Comment (raybjork) [inline spatialmath/capsule.go:280]: this is hard to follow.  Looks like youre getting the plane bisecting the capsule here?  could you add a comment?

File: spatialmath/capsule.go [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+			return false
+		}
+		if separatingAxisTest1D(centerDist, rmB.Row(i), capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+			return false
+		}
+		for j := 0; j < 3; j++ {
+			crossProductPlane := rmA.Row(i).Cross(rmB.Row(j))
+
+			// if edges are parallel, this check is already accounted for by one of the face projections, so skip this case
+			if !utils.Float64AlmostEqual(crossProductPlane.Norm(), 0, floatEpsilon) {
```
Comment (raybjork) [inline spatialmath/capsule.go:293]: I've seen this line 3 places by now, would it maybe be better to move this into the separating axis test? or is it purposefully not in there because it only gets hit on these crossProductPlanes?

File: spatialmath/capsule.go [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+				if separatingAxisTest1D(centerDist, crossProductPlane, capVec, b.halfSize, rmB) > CollisionBuffer+a.radius {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+func capsuleBoxSeparatingAxisDistance(a *capsule, b *box) float64 {
```
Comment (raybjork) [inline spatialmath/capsule.go:303]: same thing here `c *capsule`

File: spatialmath/capsule.go [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+				if separation := separatingAxisTest1D(centerDist, crossProductPlane, capVec, b.halfSize, rmB); separation > max {
+					max = separation
+				}
+			}
+		}
+	}
+	return max - a.radius
+}
+
+func separatingAxisTest1D(positionDelta, plane, capVec r3.Vector, halfSizeB [3]float64, rmB *RotationMatrix) float64 {
```
Comment (raybjork) [inline spatialmath/capsule.go:339]: is this not duplicating code?

**Review by raybjork — CHANGES_REQUESTED** [review]

File: spatialmath/capsule.go [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+
+// ToPoints converts a capsule geometry into []r3.Vector. This method takes one argument which determines
+// how many points should like on the capsule's surface. If the argument is set to 0. we automatically
+// substitute the value with defaultTotalSpherePoints.
+func (c *capsule) ToPoints(resolution float64) []r3.Vector {
+	s := &sphere{pose: NewZeroPose(), radius: c.radius}
+	vecList := s.ToPoints(resolution)
+	// move points to be correctly located on capsule endcaps
+	adj := c.length/2 - c.radius
+	for _, pt := range vecList {
```
Comment (spieswl) [inline spatialmath/capsule.go:177]: Does this block mean `vecList` has some points which are duplicated along the hemisphere of each end of the capsule?

Comment (spieswl) [reply spatialmath/capsule.go:177]: Never mind, I see what you're doing. You're splitting a sphere into its two halves and shifting the halves to each end. Nice, I like it.

File: spatialmath/capsule.go [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+		if pt.Z >= 0 {
+			pt.Z += adj
+		} else {
+			pt.Z -= adj
+		}
+	}
+
+	// Now distribute points along the cylindrical shaft
+	totalShaftPts := (c.radius * c.length) * resolution
```
Comment (spieswl) [inline spatialmath/capsule.go:186]: It doesn't look like we're using `defaultTotalSpherePoints` here (which is a little strange name-wise), but I think we need something since we could then have a `totalShaftPoints` of 0 with default sphere point densities on each end of the capsule. Does that make sense?

**Review by spieswl — CHANGES_REQUESTED** [review]
The only thing that, in my opinion, needs to be looked at is the capsule `ToPoints` resolution argument issue when being passed in a 0 value.

File: spatialmath/capsule.go [hunk 25]
```diff
@@ -0,0 +1,346 @@
...
+)
+
+// capsule is a collision geometry that represents a capsule, it has a pose and a radius that fully define it.
+//
+// ....___________________
+// .../                   \
+// .x|  |-------O-------|  |x
+// ...\___________________/
+//
+// Length is the distance between the x's, or internal segment length + 2*radius.
```
[the remark this answers is not in the document]
Comment (spieswl) [reply spatialmath/capsule.go:21]: To be clear, this is not a big deal to me.

File: spatialmath/box.go [hunk 26]
```diff
@@ -11,15 +11,51 @@ import (
 	"go.viam.com/rdk/utils"
 )
 
-// box density corresponding to how many points per square mm.
-const defaultBoxPointDensity = .5
+// Ordered list of box vertices.
+var boxVertices = [8]r3.Vector{
```
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/box.go:15]: Globals like this should generally be the first thing (or one of the first things) in a file after the imports.

I don't want to declare these in the function where they're used, because that's a performance-sensitive function and redeclaring these variables on each call will be noticeably expensive.

Perhaps a different file would be appropriate for these? `geometry_utils.go` if you like

File: spatialmath/capsule.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:21]: I put quite a bit of thought into that choice and went back and forth. This is purely a user experience decision, not something that impacts collision checks at all.

I agree that specifying segment length rather than total length makes for a "cleaner" mathematical representation, for exactly the reasons you mentioned.

However, if we do that, we make it mildly more difficult for a user to actually use. Currently the values are ones that are easy to physically measure. Let's say we go to make a capsule representation of the internal geometries of a UR5 next week. We pull out a tape measure, say "this limb is 175mm long, and 34.3mm diameter, so 17.15mm radius" and we plug that in. We don't have to worry about simple arithmetic to compute the internal segment length, which isn't something that's easily physically measured.

The majority of capsule implementations I found in the wild similarly specify full length, at least at the user-facing interfaces.

File: spatialmath/capsule.go (hunk 20 above)
Comment (biotinker) [reply spatialmath/capsule.go:293]: Not only the cross product reason, but also we avoid an entire function call.

File: spatialmath/capsule.go (hunk 22 above)
Comment (biotinker) [reply spatialmath/capsule.go:339]: Not quite; because the capsule is being modeled here as a 0x0xN box, we can only do a single `	sum -= math.Abs(capVec.Dot(plane))` rather than having to iterate through two half-sizes that are known to be zero.

A box-capsule SAT check only takes ~40ns on my computer. Adding 4 nanoseconds of computation here would mean collision checks run 10% slower. So in this case it's worth it to duplicate for performance.

Even things like passing in function pointers to specify which separating axis test function to use to reduce duplication, would be slower, because it will mess with compiler optimizations and also with branch prediction inside the CPU itself. Currently all of these loops are able to be fully unrolled and very accurately branch-predicted, which is *fast*

File: spatialmath/capsule.go (hunk 24 above)
Comment (biotinker) [reply spatialmath/capsule.go:190]: Fixed to ensure resolution is not zero

File: spatialmath/box.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (raybjork) [reply spatialmath/box.go:15]: Yeah I agree that generally you're right.  I just feel it hurts readability a bit.  You can do whatever you think is right here just wanted to raise this point

File: spatialmath/capsule.go (hunk 19 above)
Comment (biotinker) [reply spatialmath/capsule.go]: Comment was added

File: spatialmath/capsule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go:41]: I think you're right on `ParseConfig`. Was trying to save some code, but that does leave an ambiguity here that would be corrected.

Length == 0 will become an error for capsules. `ParseConfig` will look at L, and call `NewSphere` if L == 0

**Review by raybjork — APPROVED** [review]
Theres just a couple rather minor outstanding comments I left that I think are still unresolved, but after you feel you've resolved them you can merge

File: spatialmath/capsule.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (biotinker) [reply spatialmath/capsule.go]: Updated

File: spatialmath/capsule.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (spieswl) [reply spatialmath/capsule.go:21]: If you can't measure both aspects I'm not sure what the user is doing.

**Review by spieswl — APPROVED** [review]
Cool, LGTM!
